File: public/code/smc__fvg.py
"""
FVG (Fair Value Gap) Strategy
=============================
3 ardışık mumda oluşan likidite boşluğunu tespit eder, mitigation'da entry verir.

BULLISH FVG: candle[i-2].high < candle[i].low
BEARISH FVG: candle[i-2].low > candle[i].high

Entry: FVG'nin %50 mitigation seviyesinde
SL: FVG'nin ters ucu + 0.5 ATR
TP: 1:2 ve 1:3 RR
"""
import sys
import os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", ".."))

import pandas as pd
import numpy as np
from common import make_signal, atr, ema, market_structure


def detect_fvgs(df: pd.DataFrame, min_gap_atr: float = 0.3):
    """Tüm FVG'leri tespit et. Listeden döner."""
    atr_series = atr(df, 14)
    fvgs = []
    for i in range(2, len(df)):
        c1_high = df["high"].iloc[i - 2]
        c1_low = df["low"].iloc[i - 2]
        c3_high = df["high"].iloc[i]
        c3_low = df["low"].iloc[i]
        current_atr = atr_series.iloc[i]
        if pd.isna(current_atr):
            continue

        # Bullish FVG
        if c1_high < c3_low:
            gap = c3_low - c1_high
            if gap >= current_atr * min_gap_atr:
                fvgs.append({
                    "index": i,
                    "type": "bullish",
                    "top": c3_low,
                    "bottom": c1_high,
                    "midpoint": (c3_low + c1_high) / 2,
                    "filled": False,
                })

        # Bearish FVG
        elif c1_low > c3_high:
            gap = c1_low - c3_high
            if gap >= current_atr * min_gap_atr:
                fvgs.append({
                    "index": i,
                    "type": "bearish",
                    "top": c1_low,
                    "bottom": c3_high,
                    "midpoint": (c1_low + c3_high) / 2,
                    "filled": False,
                })
    return fvgs


def detect_fvg_signal(
    df: pd.DataFrame,
    min_gap_atr: float = 0.3,
    use_trend_filter: bool = True,
    ema_period: int = 200,
) -> dict:
    """
    Mevcut mumda FVG-based entry sinyali üret.

    Args:
        df: OHLCV DataFrame
        min_gap_atr: Minimum gap büyüklüğü ATR cinsinden
        use_trend_filter: HTF EMA trend filtresi
        ema_period: Trend filtresi için EMA periyodu

    Returns:
        Signal dict
    """
    if len(df) < ema_period + 5:
        return make_signal(reason="Insufficient data")

    fvgs = detect_fvgs(df, min_gap_atr)
    if not fvgs:
        return make_signal(reason="No FVG detected")

    current_price = df["close"].iloc[-1]
    current_atr = atr(df, 14).iloc[-1]
    trend_ema = ema(df["close"], ema_period).iloc[-1]
    trend = "up" if current_price > trend_ema else "down"

    # En son mitigation edilmemiş FVG'yi bul
    for fvg in reversed(fvgs):
        if fvg["filled"]:
            continue

        # Mitigation kontrolü: fiyat midpoint'e değdi mi?
        if fvg["type"] == "bullish":
            # Son 10 mumda midpoint'e değme
            recent = df.iloc[fvg["index"] + 1:]
            if len(recent) == 0:
                continue
            touched = (recent["low"] <= fvg["midpoint"]).any()
            if not touched:
                continue
            # Trend filtresi
            if use_trend_filter and trend != "up":
                continue
            # Entry koşulu: şu an midpoint civarında ve henüz tamamen dolmamış
            if current_price > fvg["bottom"] and current_price < fvg["top"]:
                entry = fvg["midpoint"]
                stop_loss = fvg["bottom"] - (0.5 * current_atr)
                risk = entry - stop_loss
                tp1 = entry + (risk * 2)
                tp2 = entry + (risk * 3)
                tp3 = entry + (risk * 5)
                return make_signal(
                    signal="long",
                    entry=entry,
                    stop_loss=stop_loss,
                    take_profit=[tp1, tp2, tp3],
                    confidence=0.75 if use_trend_filter else 0.6,
                    reason=f"Bullish FVG mitigation at {fvg['midpoint']:.4f}",
                    timestamp=int(df.index[-1].timestamp()) if hasattr(df.index[-1], "timestamp") else None,
                    metadata={"fvg": fvg, "trend": trend},
                )

        else:  # bearish
            recent = df.iloc[fvg["index"] + 1:]
            if len(recent) == 0:
                continue
            touched = (recent["high"] >= fvg["midpoint"]).any()
            if not touched:
                continue
            if use_trend_filter and trend != "down":
                continue
            if current_price > fvg["bottom"] and current_price < fvg["top"]:
                entry = fvg["midpoint"]
                stop_loss = fvg["top"] + (0.5 * current_atr)
                risk = stop_loss - entry
                tp1 = entry - (risk * 2)
                tp2 = entry - (risk * 3)
                tp3 = entry - (risk * 5)
                return make_signal(
                    signal="short",
                    entry=entry,
                    stop_loss=stop_loss,
                    take_profit=[tp1, tp2, tp3],
                    confidence=0.75 if use_trend_filter else 0.6,
                    reason=f"Bearish FVG mitigation at {fvg['midpoint']:.4f}",
                    timestamp=int(df.index[-1].timestamp()) if hasattr(df.index[-1], "timestamp") else None,
                    metadata={"fvg": fvg, "trend": trend},
                )

    return make_signal(reason="No active FVG entry")


if __name__ == "__main__":
    # Test
    np.random.seed(42)
    dates = pd.date_range("2024-01-01", periods=300, freq="15min")
    close = 100 + np.cumsum(np.random.randn(300) * 0.5)
    df = pd.DataFrame({
        "open": close + np.random.randn(300) * 0.1,
        "high": close + abs(np.random.randn(300) * 0.3),
        "low": close - abs(np.random.randn(300) * 0.3),
        "close": close,
        "volume": np.random.randint(1000, 5000, 300),
    }, index=dates)
    signal = detect_fvg_signal(df)
    print(signal)

File: public/code/smc__fvg.ts
/**
 * FVG (Fair Value Gap) Strategy
 * =============================
<<<<<<< HEAD
 * Detects 3-candle liquidity gap, entry on mitigation.
=======
 * 3 ardışık mumda likidite boşluğu tespiti, mitigation entry.
>>>>>>> 40b8debf6aee9c31feaea4d0f6fbe1f5b8d83814
 *
 * BULLISH FVG: candle[i-2].high < candle[i].low
 * BEARISH FVG: candle[i-2].low > candle[i].high
 */
import { Candle, Signal, makeSignal, atr, ema } from "../../common";

export interface FVG {
  index: number;
  type: "bullish" | "bearish";
  top: number;
  bottom: number;
  midpoint: number;
  filled: boolean;
}

export function detectFVGs(candles: Candle[], minGapAtr = 0.3): FVG[] {
  const atrValues = atr(candles, 14);
  const fvgs: FVG[] = [];

  for (let i = 2; i < candles.length; i++) {
    const c1 = candles[i - 2];
    const c3 = candles[i];
    const currentAtr = atrValues[i];
    if (isNaN(currentAtr)) continue;

    // Bullish FVG
    if (c1.high < c3.low) {
      const gap = c3.low - c1.high;
      if (gap >= currentAtr * minGapAtr) {
        fvgs.push({
          index: i,
          type: "bullish",
          top: c3.low,
          bottom: c1.high,
          midpoint: (c3.low + c1.high) / 2,
          filled: false,
        });
      }
    }
    // Bearish FVG
    else if (c1.low > c3.high) {
      const gap = c1.low - c3.high;
      if (gap >= currentAtr * minGapAtr) {
        fvgs.push({
          index: i,
          type: "bearish",
          top: c1.low,
          bottom: c3.high,
          midpoint: (c1.low + c3.high) / 2,
          filled: false,
        });
      }
    }
  }
  return fvgs;
}

export interface FVGOptions {
  minGapAtr?: number;
  useTrendFilter?: boolean;
  emaPeriod?: number;
}

export function detectFVGSignal(candles: Candle[], opts: FVGOptions = {}): Signal {
  const { minGapAtr = 0.3, useTrendFilter = true, emaPeriod = 200 } = opts;

  if (candles.length < emaPeriod + 5) {
    return makeSignal({ reason: "Insufficient data" });
  }

  const fvgs = detectFVGs(candles, minGapAtr);
  if (fvgs.length === 0) {
    return makeSignal({ reason: "No FVG detected" });
  }

  const closes = candles.map((c) => c.close);
  const currentPrice = closes[closes.length - 1];
  const atrValues = atr(candles, 14);
  const currentAtr = atrValues[atrValues.length - 1];
  const emaValues = ema(closes, emaPeriod);
  const trendEma = emaValues[emaValues.length - 1];
  const trend = currentPrice > trendEma ? "up" : "down";

  for (let i = fvgs.length - 1; i >= 0; i--) {
    const fvg = fvgs[i];
    if (fvg.filled) continue;

    if (fvg.type === "bullish") {
      const recent = candles.slice(fvg.index + 1);
      if (recent.length === 0) continue;
      const touched = recent.some((c) => c.low <= fvg.midpoint);
      if (!touched) continue;
      if (useTrendFilter && trend !== "up") continue;
      if (currentPrice > fvg.bottom && currentPrice < fvg.top) {
        const entry = fvg.midpoint;
        const stopLoss = fvg.bottom - 0.5 * currentAtr;
        const risk = entry - stopLoss;
        return makeSignal({
          signal: "long",
          entry,
          stop_loss: stopLoss,
          take_profit: [entry + risk * 2, entry + risk * 3, entry + risk * 5],
          confidence: useTrendFilter ? 0.75 : 0.6,
          reason: `Bullish FVG mitigation at ${fvg.midpoint.toFixed(4)}`,
          timestamp: candles[candles.length - 1].timestamp,
          metadata: { fvg, trend },
        });
      }
    } else {
      const recent = candles.slice(fvg.index + 1);
      if (recent.length === 0) continue;
      const touched = recent.some((c) => c.high >= fvg.midpoint);
      if (!touched) continue;
      if (useTrendFilter && trend !== "down") continue;
      if (currentPrice > fvg.bottom && currentPrice < fvg.top) {
        const entry = fvg.midpoint;
        const stopLoss = fvg.top + 0.5 * currentAtr;
        const risk = stopLoss - entry;
        return makeSignal({
          signal: "short",
          entry,
          stop_loss: stopLoss,
          take_profit: [entry - risk * 2, entry - risk * 3, entry - risk * 5],
          confidence: useTrendFilter ? 0.75 : 0.6,
          reason: `Bearish FVG mitigation at ${fvg.midpoint.toFixed(4)}`,
          timestamp: candles[candles.length - 1].timestamp,
          metadata: { fvg, trend },
        });
      }
    }
  }

  return makeSignal({ reason: "No active FVG entry" });
}
